### PR TITLE
Improve core interface

### PR DESCRIPTION
**Pull request by oeway**

### CODE REVIEW

File: imjoy/core/__init__.py [hunk 1]
```diff
@@ -25,18 +56,61 @@ class VisibilityEnum(str, Enum):
...
+    type: str
+    visibility: VisibilityEnum = VisibilityEnum.protected
+    _provider: DynamicPlugin = PrivateAttr(default_factory=lambda: None)
+
+    class Config:
+        """Set the config for pydantic."""
+
+        extra = Extra.allow
+
+    def set_provider(self, provider: DynamicPlugin) -> None:
```
Comment (MartinHjelmare) [inline imjoy/core/__init__.py:81]: Note that I've added some methods to the models when we need to access private attributes.

Comment (oeway) [reply imjoy/core/__init__.py:81]: Looks good!

File: imjoy/core/__init__.py [hunk 2]
```diff
@@ -25,18 +56,61 @@ class VisibilityEnum(str, Enum):
...
     parent: Optional[str]
     scopes: Optional[List[str]]  # a list of workspace
     expires_at: Optional[int]
+    _metadata: Dict[str, Any] = PrivateAttr(
+        default_factory=lambda: {}
+    )  # e.g. s3 credential
     _plugins: Dict[str, Any] = PrivateAttr(default_factory=lambda: {})  # id:plugin
     _sessions: List[str] = PrivateAttr(default_factory=lambda: [])  # session ids
 
+    def metadata(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (MartinHjelmare) [reply imjoy/core/__init__.py:102]: Yes!

**Comment by oeway** [comment]
Hi, I just did a quick search for `_plugins`, it seems there are still some private variable access for user_info in the server.py and interface.py. I also wonder why the linter didn't complain it.

**Pull request merged**